csi/src/node.rs

```diff
@@ -298,31 +296,74 @@ impl node_server::Node for Node {
             ));
         }
 
-        // target path will have been created previously in node_publish_volume
-        // and is one of
-        //  1. a directory for filesystem volumes ,
-        //  2. a block special file for block volumes.
-        //
-        // If it does not exist, then a previously unpublish request has
-        // succeeded.
-        let target_path = Path::new(&msg.target_path);
-        if target_path.exists() {
-            if target_path.is_dir() {
-                unpublish_fs_volume(&msg)?;
-            } else {
-                if target_path.is_file() {
-                    return Err(Status::new(
-                        Code::Unknown,
-                        format!(
-                            "Failed to unpublish volume {}: {} is a file.",
-                            &msg.volume_id, &msg.target_path
-                        ),
+        if mount::find_mount(None, Some(&msg.target_path)).is_some() {
+            debug!("Unmounting {}", msg.target_path);
+            if let Err(error) = mount::bind_unmount(&msg.target_path) {
+                return Err(failure!(
+                    Code::Internal,
+                    "Failed to unpublish volume {}: failed to unmount {}: {}",
+                    msg.volume_id,
+                    msg.target_path,
+                    error
+                ));
+            }
+        }
+
+        let mount_point = Path::new(&msg.target_path);
+        match fs::metadata(mount_point) {
+            Ok(metadata) => {
+                if metadata.is_dir() {
+                    // Mount Volume
+                    debug!("Removing directory {}", msg.target_path);
+                    if let Err(error) = fs::remove_dir(mount_point) {
+                        return Err(failure!(
+                            Code::Internal,
```

Comment:
I guess there's not a more appropriate error code than `Internal` for this?

Reply:
Hadn't really given it too much thought as that was essentially what was used already, but no there is not really a better alternative.